### ISSUE TITLE
example problem not working: NIST_ruby/multiruby_dexelas/ 

### ISSUE THREAD

**Issue by donald-e-boyce** [HEXRD/hexrd#731]
I tried to run fit-grains on the `mruby_config_composite.yml` example in `NIST_ruby/multiruby_dexelas/ ` and got a
```TypeError: cannot pickle '_thread.lock' object``` error message.

This was run using the latest hexrd: commit ba52e1a1bc614ad289b23afa7f7db27878873a1a from 11/15.

```(hexrd-dev) (MacBook-Pro-2023: run) 1027. !hexrd
hexrd fit-grains mruby_config_composite.yml 
24-11-22 15:02:41 - "working_dir" not specified, defaulting to "/Users/deboyce/Source/hexrd/examples/NIST_ruby/multiruby_dexelas/run"
24-11-22 15:02:41 - material:reset_exclusions not specified, defaulting to True
24-11-22 15:02:41 - material:dmax not specified, defaulting to None
24-11-22 15:02:41 - material:tthmin not specified, defaulting to None
24-11-22 15:02:41 - material:tthmax not specified, defaulting to None
24-11-22 15:02:41 - material:sfacmin not specified, defaulting to 0.05
24-11-22 15:02:41 - material:sfacmax not specified, defaulting to None
24-11-22 15:02:41 - material:pintmin not specified, defaulting to None
24-11-22 15:02:41 - material:pintmax not specified, defaulting to None
24-11-22 15:02:41 - fit_grains:min_sfac_ratio not specified, defaulting to None
24-11-22 15:02:41 - fit_grains:reset_exclusions not specified, defaulting to True
24-11-22 15:02:41 - fit_grains:dmin not specified, defaulting to None
24-11-22 15:02:41 - fit_grains:dmax not specified, defaulting to None
24-11-22 15:02:41 - fit_grains:tthmin not specified, defaulting to None
24-11-22 15:02:41 - fit_grains:tthmax not specified, defaulting to None
24-11-22 15:02:41 - fit_grains:sfacmin not specified, defaulting to None
24-11-22 15:02:41 - fit_grains:sfacmax not specified, defaulting to None
24-11-22 15:02:41 - fit_grains:pintmin not specified, defaulting to None
24-11-22 15:02:41 - fit_grains:pintmax not specified, defaulting to None
24-11-22 15:02:41 - fit_grains:estimate not specified, defaulting to None
24-11-22 15:02:41 - Missing /Users/deboyce/Source/hexrd/examples/NIST_ruby/multiruby_dexelas/run/accepted_orientations_composite_py310_ruby.dat, running find-orientations
omega period specification is deprecated
24-11-22 15:02:53 - === begin fit-grains ===
24-11-22 15:02:54 - using 74 HKLs
24-11-22 15:02:54 - *** begin analysis "composite_py310" ***
24-11-22 15:02:54 - logging to /Users/deboyce/Source/hexrd/examples/NIST_ruby/multiruby_dexelas/run/composite_py310/fit-grains.log
24-11-22 15:02:54 - fit_grains:estimate not specified, defaulting to None
24-11-22 15:02:54 - 	starting fit on 3 processes with chunksize 1
Traceback (most recent call last):
  File "/Users/deboyce/miniconda3/envs/hexrd-dev/bin/hexrd", line 8, in <module>
    sys.exit(main())
  File "/Users/deboyce/Source/hexrd/pkg/hexrd/cli/main.py", line 85, in main
    args.func(args, p)
  File "/Users/deboyce/Source/hexrd/pkg/hexrd/cli/fit_grains.py", line 339, in execute
    fit_results = fit_grains(
  File "/Users/deboyce/Source/hexrd/pkg/hexrd/fitgrains.py", line 384, in fit_grains
    pool = multiprocessing.Pool(
  File "/Users/deboyce/miniconda3/envs/hexrd-dev/lib/python3.9/multiprocessing/context.py", line 119, in Pool
    return Pool(processes, initializer, initargs, maxtasksperchild,
  File "/Users/deboyce/miniconda3/envs/hexrd-dev/lib/python3.9/multiprocessing/pool.py", line 212, in __init__
    self._repopulate_pool()
  File "/Users/deboyce/miniconda3/envs/hexrd-dev/lib/python3.9/multiprocessing/pool.py", line 303, in _repopulate_pool
    return self._repopulate_pool_static(self._ctx, self.Process,
  File "/Users/deboyce/miniconda3/envs/hexrd-dev/lib/python3.9/multiprocessing/pool.py", line 326, in _repopulate_pool_static
    w.start()
  File "/Users/deboyce/miniconda3/envs/hexrd-dev/lib/python3.9/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
  File "/Users/deboyce/miniconda3/envs/hexrd-dev/lib/python3.9/multiprocessing/context.py", line 284, in _Popen
    return Popen(process_obj)
  File "/Users/deboyce/miniconda3/envs/hexrd-dev/lib/python3.9/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/Users/deboyce/miniconda3/envs/hexrd-dev/lib/python3.9/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/Users/deboyce/miniconda3/envs/hexrd-dev/lib/python3.9/multiprocessing/popen_spawn_posix.py", line 47, in _launch
    reduction.dump(process_obj, fp)
  File "/Users/deboyce/miniconda3/envs/hexrd-dev/lib/python3.9/multiprocessing/reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
TypeError: cannot pickle '_thread.lock' object
```

**Comment by psavery** [comment]
I wonder if this might be a mac-specific problem. I was able to run it on Linux.

**Comment by donald-e-boyce** [comment]
I can try it on linux too.

**Comment by psavery** [comment]
I'll try it on Mac next week.

**Comment by psavery** [comment]
@donald-e-boyce I was able to reproduce this on Mac. I guess our tests did not catch this. I think our fit grains test runs on the single grain GE example (which wouldn't have multiprocessing), and that might be why it wasn't caught.

Looks like it is caused by the fact that we are multiprocessing across grains during fit-grains, the imageseries is being sent to the different processes via pickling, and [this recently added lock](https://github.com/HEXRD/hexrd/blob/ba52e1a1bc614ad289b23afa7f7db27878873a1a/hexrd/imageseries/load/framecache.py#L35) is not pickle-able.

If you want to run it right now, you might want to just disable the multiprocessing in the config options (set it to `1`). We can fix this next week.

We had ran into a similar problem with pickling HDF5 imageseries in the past, and it was fixed by adding [these attributes](https://github.com/HEXRD/hexrd/blob/ba52e1a1bc614ad289b23afa7f7db27878873a1a/hexrd/imageseries/load/hdf5.py#L80-L104), which define what happens during pickling and unpickling. I think we'll just need to add similar functions for the frame-cache imageseries, to avoid pickling that lock.
